### PR TITLE
CARD4L Optical Extension (SR + ST)

### PR DESCRIPTION
**Pull request by m-mohr**
### The STAC CARD4L Optical Extension has been migrated to a separate repository: <https://github.com/stac-extensions/card4l/tree/main/optical>

### CODE REVIEW

**Comment by sophieherrmann** [comment]
How would per pixel metadata like `No Data`, `Incomplete Testing`, `Saturation`, `Cloud` and `Cloud Shadow` be added?
If I got it correct all of them are minimum requirements, or?

I assume this information would probably be stored in separate files. Therefore maybe adding them in the `Assets` could be an option?

**Comment by m-mohr** [comment]
@sophieherrmann This extension is - unfortunately - incomplete and the per pixel metadata is missing, see the ToDos in the file. I'll try to complete it tomorrow with help of Peter. You are right that it will likely be separate files, similarly how it's done in the SAR extension. Sorry for the inconvenience.

**Comment by sophieherrmann** [comment]
Ah yes I missed the ToDos. But the `General Metadata` part seems to be complete, or?

**Comment by m-mohr** [comment]
@sophieherrmann We've realized today that some assets can also be handled as general metadata, which may add some additional fields there, but other than that it should be mostly complete.

File: extensions/card4l-eo/README.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+### STAC Item Properties
+
+#### CARD4L
+
+| Field Name                         | Data Type | Description                                                  | Req.  |
+| ---------------------------------- | --------- | ------------------------------------------------------------ | ----- |
+| card4l:specification               | string    | **REQUIRED.** The CARD4L specification implemented, either `SR` (Optical, Surface Reflectance) or `ST` (Optical, Surface Temperature). | *n/a* |
+| card4l:specification_version       | string    | **REQUIRED.** The CARD4L specification version. Currently always `5.0`. | *n/a* |
+| card4l:northern_geometric_accuracy | number    | An estimate of the northern geometric accuracy in meters.    | 1.8   |
```
Comment (matthewhanson) [inline extensions/card4l-eo/README.md:53]: Previously discussed with @m-mohr , but geometric accuracy I think is more generally applicable outside of CARD4L.  It may make sense to have an extension that includes geometric and radiometric accuracies.  Does it fit in with the `processing` extension?

File: extensions/card4l-eo/README.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+| -------------- | ------------------------------------------------------------ | ------ |
+| eo:cloud_cover |                                                              | 1.17   |
+| eo:bands       | You may include all bands specified in the assets again as an overview. | (1.10) |
+
+#### Processing
+
+| Field Name          | Description                                                  | Req.        |
+| ------------------- | ------------------------------------------------------------ | ----------- |
+| processing:facility |                                                              | 1.15        |
+| processing:software | String likely needs to be split into software name and version number. | 1.13 / 1.15 |
```
Comment (matthewhanson) [inline extensions/card4l-eo/README.md:82]: This doesn't seem very specific, does it need to include software and version or is it more or less a free text field?

File: extensions/card4l-eo/README.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+| processing:lineage  | Additional processing information and parameters, e.g. a description of the processing chain. | 1.15        |
```
Comment (matthewhanson) [inline extensions/card4l-eo/README.md:83]: These tables don't include `Type` like in the main spec and in extensions. Is it a string? Is there any guidance on what to put here or just a prose description of the processing?

File: extensions/card4l-eo/README.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+| Field Name           | Description                                                  | Req. |
+| -------------------- | ------------------------------------------------------------ | ---- |
+| view:off_nadir       | **REQUIRED.** The average off-nadir angle, for per-pixel angles, refer to the asset with the key `off-nadir`. Convert to degree, if required. | \*   |
+| view:incidence_angle | **REQUIRED.** The average incidence angle, for per-pixel angles, refer to the asset with the key `incidence-angle`. Convert to degree, if required. | \*   |
+| view:azimuth         | **REQUIRED.** The average azimuth angle, for per-pixel angles, refer to the asset with the key `azimuth`. Convert to degree, if required. | \*   |
+| view:sun_azimuth     | **REQUIRED.** The average sun azimuth angle, for per-pixel angles, refer to the asset with the key `sun-azimuth`. Convert to degree, if required. | \*   |
+| view:sun_elevation   | **REQUIRED.** The average sub elevation angle, for per-pixel angles, refer to the asset with the key `sun-elevation`. Convert to degree, if required. | \*   |
+
+\* = Requirement 2.8 for Surface Temperature (ST) / 2.11 for Surface Reflectance (SR)
```
Comment (matthewhanson) [inline extensions/card4l-eo/README.md:103]: Does this mean sun elevation and azimuth are required for Surface Temperature?
Sun illumination angles don't factor into temperature

File: extensions/card4l-eo/README.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+### STAC Item Links
+
+| Relation Type            | Description                                                  | Req.                |
+| ------------------------ | ------------------------------------------------------------ | ------------------- |
+| card4l-document          | **REQUIRED.** Provides at least one link to the CARD4L specification document. Word (media type: `application/vnd.openxmlformats-officedocument.wordprocessingml.document`) and/or PDF (media type: `application/pdf`). | *n/a*               |
+| derived_from             | Points back to the source's STAC Item. May be multiple items, if the product is derived from multiple acquisitions. | 1.15                |
+| about                    | Link to algorithms used in the generation process. See also the [notes](#notes) regarding req. 1.13. | 1.13                |
+| related                  | **REQUIRED.** Link to the sources of ancillary or auxiliary data used in the generation process. Excludes DEMs, which use the relation `elevation-model` instead. | 1.14                |
```
Comment (matthewhanson) [inline extensions/card4l-eo/README.md:112]: What does it mean to link to the data source? Is it a URL to a file, or a website landing page indicating how to download the data?
This excludes DEMs, so I think some examples of ancillary data would be useful, e.g., Ozone, Water Vapor

File: extensions/card4l-eo/README.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+| access                   | STRONGLY RECOMMENDED. Link to data access information.       | 1.16                |
```
Comment (matthewhanson) [inline extensions/card4l-eo/README.md:113]: So, a landing page to this dataset?
Seems more useful at the Collection level

File: extensions/card4l-eo/README.md (hunk 5 above)
Comment (matthewhanson) [reply extensions/card4l-eo/README.md:112]: Now I see ozone, water vapor, emissivity and others are all their own relation types.  So since this is required I think an example of what this would include should be here.

File: extensions/card4l-eo/README.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+### STAC Item Assets
+
+Whether the metadata are provided in a single record relevant to all pixels, or separately for each pixel, is at the discretion of the data provider. 
+
+Each of the assets can either be exposed individually or grouped together in any form. In the latter case the role names can simply be merged to a set of unique role names. Roles can also be combined for a single file. For example, a cloud mask which is also including cloud shadows can use the roles `cloud` and `cloud-shadow` for a single file. The `file:values` property can the contain which value(s) correspond to clouds and which value(s) correspond to cloud shadows respectively.
+
+The italic role names are proposed to be the asset's key.
+
+| Role Name(s)                           | Additional properties                                        | Description                                                  | Req.                 |
```
Comment (matthewhanson) [inline extensions/card4l-eo/README.md:137]: This is a useful table, but it's not clear if the Additional Properties are required....I think that is the intention and from what I know of CARD4L.

File: extensions/file/README.md [hunk 8]
```diff
@@ -14,19 +14,42 @@ Provides a way to specify file related details such as checksum, data type and s
...
-| file:checksum | string | Provides a way to specify file [checksums](#checksums) (e.g. BLAKE2, MD5, SHA1, SHA2, SHA3). The hashes are self-identifying hashes as described in the [Multihash specification](https://github.com/multiformats/multihash) and must be encoded as hexadecimal (base 16) string with lowercase letters. |
-| file:data_type | string | The data type of the file. One of the [data types](#data-types) below. |
-| file:header_size | integer | The header [size](#sizes) of the file, specified in bytes. |
-| file:size | integer | The file [size](#sizes), specified in bytes. |
+The following fields can be used for Links (in the [`Link Object`](../../item-spec/item-spec.md#link-object)) and assets (in the [`Asset Object`](../../item-spec/item-spec.md#asset-object)) although most of these fields make more sense for assets and are usually not very useful for links.
+
+| Field Name           | Type                                    | Description                                                  |
+| -------------------- | --------------------------------------- | ------------------------------------------------------------ |
```
Comment (matthewhanson) [inline extensions/file/README.md:20]: Might be better to make this into a separate PR?

Comment (m-mohr) [reply extensions/file/README.md:20]: It is

File: extensions/card4l-eo/README.md (hunk 7 above)
Comment (m-mohr) [reply extensions/card4l-eo/README.md:137]: Req are given in the table below

File: extensions/card4l-eo/README.md (hunk 6 above)
Comment (m-mohr) [reply extensions/card4l-eo/README.md:113]: There's no req for having a Collection

File: extensions/card4l-eo/README.md (hunk 5 above)
Comment (m-mohr) [reply extensions/card4l-eo/README.md:112]: I have no more details as the spec hss and that should be read side by side anyway. Don't want to duplicate all the information given there

File: extensions/card4l-eo/README.md (hunk 3 above)
Comment (m-mohr) [reply extensions/card4l-eo/README.md:83]: It's 1:1 like in the main description. I have not included data type and original description to avoid duplication

File: extensions/card4l-eo/README.md (hunk 2 above)
Comment (m-mohr) [reply extensions/card4l-eo/README.md:82]: It's 1:1 like in the main extension . I have not included data type and original description to avoid duplication

File: extensions/card4l-eo/README.md (hunk 1 above)
Comment (m-mohr) [reply extensions/card4l-eo/README.md:53]: Agreed, but not sure about the new place. It's also included in Card4l Sar

**Comment by sophieherrmann** [comment]
I think `view:off_nadir` is not a minimum requirement at least for SR, or? CARD4L 2.11 states:

> Provide average solar and sensor viewing azimuth and zenith angles.

Which should be covered with `view:azimuth`, `view:incidence_angle`, `view:sun_azimuth` and `view:sun_elevation`.

**Comment by m-mohr** [comment]
@sophieherrmann Yes, that seems to be correct. I've removed the requirement and asset role.

File: extensions/card4l-eo/README.md (hunk 4 above)
Comment (m-mohr) [reply extensions/card4l-eo/README.md:103]: Yes, they are. Both specs share the same requirement:

> Provide average solar and sensor viewing azimuth and zenith angles.

File: extensions/card4l-eo/README.md (hunk 3 above)
Comment (m-mohr) [reply extensions/card4l-eo/README.md:83]: I've added the following part at the top of the extension:

> You have to read the STAC extensions in combination with this extension as this extension just provides the mapping between the STAC fields and the CARD4L requirements, but this extension doesn't give information on the data type or an actual detailed description about the fields.

File: extensions/card4l-eo/README.md (hunk 2 above)
Comment (m-mohr) [reply extensions/card4l-eo/README.md:82]: see https://github.com/radiantearth/stac-spec/pull/933#discussion_r569345666

File: extensions/card4l-eo/README.md (hunk 6 above)
Comment (m-mohr) [reply extensions/card4l-eo/README.md:113]: I've added the following chapter to the spec, which should help users to de-duplicate assets and links:

## STAC Collections

CARD4L lists a lot of requirements (and fields) that have common values across all generated STAC Items and assets.
Thus, it is **recommended** to provide a STAC Collection for the Items and put common fields (in the STAC Item `properties`)
into [Collection `summaries`](../../collection-spec/collection-spec.md#collection-fields).
While the STAC Item fields still need to be in the Item, too, you can de-duplicate links and assets by putting common
links once into the STAC Collection links. Also, common assets can be just put once into the STAC Collection using the
STAC extension [Collection Assets](../collection-assets/README.md).

File: extensions/card4l-eo/README.md (hunk 7 above)
Comment (m-mohr) [reply extensions/card4l-eo/README.md:137]: Required properties are now bold, too.

**Comment by m-mohr** [comment]
The STAC CARD4L Optical Extension has been migrated to a separate repository: https://github.com/stac-extensions/card4l/tree/main/optical
